lib/cn/hblock_builder.c

```diff
@@ -0,0 +1,327 @@
+/* SPDX-License-Identifier: Apache-2.0 */
+/*
+ * Copyright (C) 2022 Micron Technology, Inc.  All rights reserved.
+ */
+
+ #define MTF_MOCK_IMPL_hblock_builder
+
+#include <errno.h>
+#include <stdlib.h>
+
+#include <hse/limits.h>
+
+#include <hse_ikvdb/blk_list.h>
+#include <hse_ikvdb/cn.h>
+#include <hse_ikvdb/kvset_builder.h>
+#include <hse_ikvdb/mclass_policy.h>
+#include <hse_util/alloc.h>
+#include <hse_util/event_counter.h>
+#include <hse_util/hse_err.h>
+#include <hse_util/hlog.h>
+#include <hse_util/page.h>
+#include <hse_util/storage.h>
+
+#include "hblock_builder.h"
+#include "omf.h"
+#include "wbt_builder.h"
+#include "blk_list.h"
+
+struct hblock_builder {
+    struct mpool *mpool;
+    const struct cn *cn;
+    struct wbb *ptree;
+    unsigned int ptree_pgc;
+    uint32_t max_size;
+    enum hse_mclass_policy_age agegroup;
+};
+
+static unsigned int
+available_pgc(struct hblock_builder *bld)
+{
+    return (bld->max_size / PAGE_SIZE) - HBLOCK_HDR_PAGES - HLOG_PGC;
+}
+
+static void HSE_NONNULL(1)
+make_header(
+    struct hblock_hdr_omf *hdr,
+    const uint64_t min_seqno,
+    const uint64_t max_seqno,
+    const uint32_t num_kblocks,
+    const uint32_t num_vblocks,
+    const uint32_t num_vgroups,
+    const uint32_t ptree_pgc,
+    const uint32_t vblk_idx_adj_pgc,
+    const struct key_obj *const min_pfx,
+    const struct key_obj *const max_pfx)
+{
+    /* If we have one, we must have both */
+    assert(((min_pfx && max_pfx) || (!min_pfx && !max_pfx)));
+
+    omf_set_hbh_magic(hdr, HBLOCK_HDR_MAGIC);
+    omf_set_hbh_version(hdr, HBLOCK_HDR_VERSION);
+    omf_set_hbh_min_seqno(hdr, min_seqno);
+    omf_set_hbh_max_seqno(hdr, max_seqno);
+    omf_set_hbh_num_kblocks(hdr, num_kblocks);
+    omf_set_hbh_num_vblocks(hdr, num_vblocks);
+    omf_set_hbh_num_vgroups(hdr, num_vgroups);
+    omf_set_hbh_hlog_off_pg(hdr, HBLOCK_HDR_PAGES);
+    omf_set_hbh_hlog_len_pg(hdr, HLOG_PGC);
+    omf_set_hbh_ptree_data_off_pg(hdr, HBLOCK_HDR_PAGES + HLOG_PGC);
+    omf_set_hbh_ptree_data_len_pg(hdr, ptree_pgc);
+    omf_set_hbh_vblk_idx_adj_off_pg(hdr, HBLOCK_HDR_PAGES + HLOG_PGC + ptree_pgc);
+    omf_set_hbh_vblk_idx_adj_len_pg(hdr, vblk_idx_adj_pgc);
+
+    if (max_pfx) {
+        unsigned int max_pfx_len = 0;
+
+        /* Make sure there is enough room in the space allocated */
+        static_assert(BYTE_ALIGN(HBLOCK_HDR_LEN - 2 * HSE_KVS_PFX_LEN_MAX) >= sizeof(*hdr),
```

Comment:
This static_assert assumes that HBLOCK_HDR_LEN includes 2 * HSE_KVS_P{FX_LEN_MAX. Please see my other comment in omf.h